File: proposed-projects/round-trip-experiment/reverse/20260909T182714Z_2643e4be/20260910T100600Z_9e4cc6a6/ported_implementation/gbnf/rules_builder/rules_builder_types.py
"""
Internal (linear) rule definitions produced by the RulesBuilder, before they
are folded into a stack of paths by `build_rule_stack`.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Union


class InternalBase:
    def __eq__(self, other: object) -> bool:
        return isinstance(other, type(self))

    def __hash__(self) -> int:
        return id(self)

    def __repr__(self) -> str:
        return f"{type(self).__name__}()"


class InternalBaseWithValue(InternalBase):
    def __init__(self, value: Any) -> None:
        self.value = value

    def __eq__(self, other: object) -> bool:
        return isinstance(other, type(self)) and other.value == self.value

    def __hash__(self) -> int:
        return id(self)

    def __repr__(self) -> str:
        return f"{type(self).__name__}(value={self.value})"


class InternalBaseWithInt(InternalBaseWithValue):
    def __init__(self, value: int) -> None:
        super().__init__(value)


class InternalBaseWithListOfInts(InternalBaseWithValue):
    def __init__(self, value: Optional[List[int]] = None) -> None:
        # copy the incoming list
        super().__init__(list(value) if value is not None else [])


class InternalRuleDefWithNumericValue(InternalBaseWithInt):
    pass


class InternalRuleDefChar(InternalBaseWithListOfInts):
    pass


class InternalRuleDefCharAlt(InternalBaseWithInt):
    pass


class InternalRuleDefAlt(InternalBase):
    pass


class InternalRuleDefCharNot(InternalBaseWithListOfInts):
    pass


class InternalRuleDefCharRngUpper(InternalBaseWithInt):
    pass


class InternalRuleDefReference(InternalBaseWithInt):
    pass


class InternalRuleDefEnd(InternalBase):
    pass


class InternalRuleDefWithoutValue(InternalBase):
    pass


InternalRuleDef = Union[
    InternalRuleDefChar,
    InternalRuleDefEnd,
    InternalRuleDefReference,
    InternalRuleDefCharNot,
    InternalRuleDefWithNumericValue,
    InternalRuleDefWithoutValue,
    InternalRuleDefAlt,
    InternalRuleDefCharRngUpper,
    InternalRuleDefCharAlt,
]

InternalRuleDefCharOrAltChar = Union[InternalRuleDefChar, InternalRuleDefCharAlt]

SymbolIds = Dict[str, int]


def is_rule_def_alt(rule: object) -> bool:
    return isinstance(rule, InternalRuleDefAlt)


def is_rule_def_ref(rule: object) -> bool:
    return isinstance(rule, InternalRuleDefReference)


def is_rule_def_end(rule: object) -> bool:
    return isinstance(rule, InternalRuleDefEnd)


def is_rule_def_char(rule: object) -> bool:
    return isinstance(rule, InternalRuleDefChar)


def is_rule_def_char_not(rule: object) -> bool:
    return isinstance(rule, InternalRuleDefCharNot)


def is_rule_def_char_alt(rule: object) -> bool:
    return isinstance(rule, InternalRuleDefCharAlt)


def is_rule_def_char_rng_upper(rule: object) -> bool:
    return isinstance(rule, InternalRuleDefCharRngUpper)

File: packages/gbnf-experiment/data/20260909T182714Z_2643e4be/ported_implementation/src/rules-builder/rules-builder-types.ts
/**
 * Internal (linear) rule definitions produced by the RulesBuilder, before they
 * are folded into a stack of paths by `buildRuleStack`.
 */

export class InternalBase {
  equals(other: unknown): boolean {
    return other instanceof this.constructor;
  }
}

export class InternalBaseWithValue<T = unknown> extends InternalBase {
  value: T;

  constructor(value: T) {
    super();
    this.value = value;
  }

  override equals(other: unknown): boolean {
    return (
      other instanceof this.constructor &&
      JSON.stringify((other as InternalBaseWithValue<T>).value) ===
        JSON.stringify(this.value)
    );
  }
}

export class InternalBaseWithInt extends InternalBaseWithValue<number> {}

export class InternalBaseWithListOfInts extends InternalBaseWithValue<number[]> {
  constructor(value: number[] = []) {
    // the reference implementation copies the incoming list
    super([...value]);
  }
}

export class InternalRuleDefWithNumericValue extends InternalBaseWithInt {}

export class InternalRuleDefChar extends InternalBaseWithListOfInts {}

export class InternalRuleDefCharAlt extends InternalBaseWithInt {}

export class InternalRuleDefAlt extends InternalBase {}

export class InternalRuleDefCharNot extends InternalBaseWithListOfInts {}

export class InternalRuleDefCharRngUpper extends InternalBaseWithInt {}

export class InternalRuleDefReference extends InternalBaseWithInt {}

export class InternalRuleDefEnd extends InternalBase {}

export class InternalRuleDefWithoutValue extends InternalBase {}

export type InternalRuleDef =
  | InternalRuleDefChar
  | InternalRuleDefEnd
  | InternalRuleDefReference
  | InternalRuleDefCharNot
  | InternalRuleDefWithNumericValue
  | InternalRuleDefWithoutValue
  | InternalRuleDefAlt
  | InternalRuleDefCharRngUpper
  | InternalRuleDefCharAlt;

export type InternalRuleDefCharOrAltChar =
  | InternalRuleDefChar
  | InternalRuleDefCharAlt;

export type SymbolIds = Record<string, number>;

export const isRuleDefAlt = (rule: unknown): rule is InternalRuleDefAlt =>
  rule instanceof InternalRuleDefAlt;

export const isRuleDefRef = (rule: unknown): rule is InternalRuleDefReference =>
  rule instanceof InternalRuleDefReference;

export const isRuleDefEnd = (rule: unknown): rule is InternalRuleDefEnd =>
  rule instanceof InternalRuleDefEnd;

export const isRuleDefChar = (rule: unknown): rule is InternalRuleDefChar =>
  rule instanceof InternalRuleDefChar;

export const isRuleDefCharNot = (
  rule: unknown,
): rule is InternalRuleDefCharNot => rule instanceof InternalRuleDefCharNot;

export const isRuleDefCharAlt = (
  rule: unknown,
): rule is InternalRuleDefCharAlt => rule instanceof InternalRuleDefCharAlt;

export const isRuleDefCharRngUpper = (
  rule: unknown,
): rule is InternalRuleDefCharRngUpper =>
  rule instanceof InternalRuleDefCharRngUpper;
